exchange/exchange.go

```diff
@@ -305,6 +312,131 @@ func (e *exchange) buildBidResponse(ctx context.Context, liveAdapters []openrtb_
 	return bidResponse, err
 }
 
+func applyCategoryMapping(requestExt openrtb_ext.ExtRequest, seatBids *map[openrtb_ext.BidderName]*pbsOrtbSeatBid, categoriesFetcher stored_requests.CategoryFetcher) (map[string]string, error) {
+	res := make(map[string]string)
+
+	//If includebrandcategory is present in ext then CE feature is on.
+	if requestExt.Prebid.Targeting == nil {
+		return res, nil
+	}
+	brandCatExt := requestExt.Prebid.Targeting.IncludeBrandCategory
+
+	//If includebrandcategory is present in ext then CE feature is on.
+	if brandCatExt == (openrtb_ext.ExtIncludeBrandCategory{}) {
+		return res, nil //if not present continue the existing processing without CE.
+	}
+
+	//if includebrandcategory present but primaryadserver/publisher not present then error out the request right away.
+	primaryAdServer, err := getPrimaryAdSetver(brandCatExt.PrimaryAdServer) //1-Freewheel 2-DFP
+	if err != nil {
+		return res, err
+	}
+
+	publisher := brandCatExt.Publisher
+
+	seatBidsToRemove := make([]openrtb_ext.BidderName, 0)
+
+	for bidderName, seatBid := range *seatBids {
+		bidsToRemove := make([]int, 0)
+		for bidInd := range seatBid.bids {
+			bid := seatBid.bids[bidInd].bid
+			var tempRespBidExt ResponseBidExt
+			if len(bid.Ext) > 0 {
+				fmt.Println(string(bid.Ext))
```

Comment:
Remove or comment out debug prints before merging to master.

Reply:
Done!